File: app/src/main/java/com/example/consultants/match/model/ContactObserver.java
package com.example.consultants.match.model;

import com.example.consultants.match.model.jsondata.ContactResponse;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;

public class ContactObserver implements Observer<ContactResponse> {

    Callback callback;

    public static ContactObserver addCallback(Callback callback) {

        ContactObserver contactObserver = new ContactObserver();
        contactObserver.setCallback(callback);
        return contactObserver;
    }

    private void setCallback(Callback callback) {
        this.callback = callback;
    }

    @Override
    public void onSubscribe(Disposable d) {

    }

    @Override
    public void onNext(ContactResponse contacts) {
        callback.onNext(contacts);
    }

    @Override
    public void onError(Throwable e) {
        callback.onError(e.getMessage());
    }

    @Override
    public void onComplete() {

    }

    public interface Callback {
        void onNext(ContactResponse contacts);
        void onError(String error);
    }
}

File: app/src/main/java/com/example/consultants/match/model/ContactObserver.kt
package com.example.consultants.match.model

import com.example.consultants.match.model.jsondata.ContactResponse
import io.reactivex.Observer
import io.reactivex.disposables.Disposable

class ContactObserver : Observer<ContactResponse> {

    internal var callback: Callback? = null

    private fun setCallback(callback: Callback) {
        this.callback = callback
    }

    override fun onSubscribe(d: Disposable) {

    }

    override fun onNext(contacts: ContactResponse) {
        callback?.onNext(contacts)
    }

    override fun onError(e: Throwable) {
        callback?.onError(e.message.toString())
    }

    override fun onComplete() {

    }

    interface Callback {
        fun onNext(contactResponse: ContactResponse)
        fun onError(error: String)
    }

    companion object {

        fun addCallback(callback: Callback): ContactObserver {

            val contactObserver = ContactObserver()
            contactObserver.setCallback(callback)
            return contactObserver
        }
    }
}
